README.md
# cmpp

cmpp v2.0 / v3.0 的实现方案

## Usage

```javascript
const Client = require('cmpp');

const client = new Client({
  host: '127.0.0.1',
  port: 7890,
  serviceId: 'serviceName',
  srcId: '10xxxxxx',
  heartbeatMaxAttempts: 3,
  heartbeatInterval: 3 * 60 * 1000,
  heartbeatTimeout: 60 * 1000,
  timeout: 30 * 1000,
  mobilesPerSecond: 200,
  version: 0x20, // or 0x30,
});

// 每秒的发送量
client.on('mobile_count', (sendingMobileCount) => {});

// 设置「上行回调」处理方法, 同步异步均支持
// 方法执行完成后发送 CMPP_DELIVER_RESP, 异常失败时 Result = 7 (业务代码错)
client.setReceive(async(body = {}) => {
  // console.log(body);
});

// 设置「下行回调」处理方法，同步异步均支持
// 方法执行完成后发送 CMPP_DELIVER_RESP, 异常失败时 Result = 7 (业务代码错)
client.setDeliver(async(body = {}) => {
  // console.log(body);
});

// 报错信息
client.on('error', (err) => {});

// 请求超时, command 为命令, 比如 CMPP_CONNECT
client.on('timeout', (command) => {});

client.on('terminated', (message) => {
  // 连接主动断开和心跳停止断开
  // undefined or 'heartbeat'
  console.log(message);
});

const SourceAddr = '100100'; // 源地址，此处为 SP_Id，即 SP 的企业代码
const SharedSecret = 'password'; // 授权密码
client.connect(SourceAddr, SharedSecret).catch(err => {});

// client.send(mobile, content, extendCode); // 扩展码可不填
client.send('1358888xxxx', '【签名】这里是内容', '1234');

// 批量发送, 上限 100 个号码
// client.sendGroup(['1358888xxxx', '1368888xxxx'], '【签名】这里是内容', '1234');

// 断开连接
client.disconnect()
```

---

## 备注

v0.x.x 基于 [@nbcat](https://www.npmjs.com/~nbnat) 实现的 [nb-cmpp-2](https://www.npmjs.com/package/nb-cmpp-2) 项目重写.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [v4.0.0] - 2021-06-11

### Changed

  - [BREAKING] rename `client.socket` to `client.conn`
  - 更新了 `send buffer` 和 `receive buffer` 的日志，输出更便于理解

### Fixed

  - 解决 socket 新旧交替的异常, 需要避免旧逻辑的 destroySocket 删除新逻辑的 socket


## [v3.0.2] - 2021-06-11

### Added
  - 优化 CMPP 鉴权锁，引入 `cmppConnected` 作为鉴权完成的标志

### Fixed
  - 要求 TCP 连接正常后发送 CMPP_CONNECT

## [v3.0.1] - 2021-05-27

### Fixed

  - 优化 Fee_terminal_Id 字段解析方式，由 number 类型改为 string 类型, 方便解析(不影响使用)

### Added

  - add cmpp protocol connection images
  - add CHANGELOG.md

## [v3.0.0] - 2021-01-21

### Changed

  - [BREAKING] 引入 `client.setReceive(async() => {})` 和 `client.setDeliver(async() => {})` 设置「回调处理方法」。由原来的**并行**处理 「`CMPP_DELIVER_RESP` 响应」 / 「回调处理方法」改为**先**执行「回调处理方法」**后**执行「`CMPP_DELIVER_RESP` 响应」
  - `CMPP_ACTIVE_TEST_RESP` 内容缺少 `Reserved`

## [v2.2.0] - 2020-12-09

### Added
  - 支持 `CMPP v3.0`

lib/client.js
const EventEmitter = require('events')
const iconv = require('iconv-lite')
const cmdCfg = require('./commandsConfig')
const Socket = require('./cmppSocket')
const isPromise = require('is-promise');

const RetryInfo = 'please retry later';

class Client extends EventEmitter {
  constructor(config) {
    super()
    this.config = config
    this.contentLimit = 70
    this.longSmsBufLimit = 140
    this.sendingMobileCount = 0
    _.defaults(config, require('./config'))

    this.debug = require('debug')(`cmpp${config.version.toString(16)}:client:${config.host}:${config.port}`)

    this.conn = new Socket(config)
    // 上行回调处理方法
    this.receive = (body) => {
      console.log('receive data', body);
    }
    // 下行回调处理方法
    this.deliver = (body) => {
      console.log('Get deliver data', body);
    }

    this.bindEvent()
    this.clearMobileCount()
  }

  connect(spId, secret) {
    const _this = this
    const timestamp = _this.getTimestamp()

    if (this.conn.isReady) {
      return Promise.resolve()
    }
    this.spId = spId

    return this.conn.connect(this.config.port, this.config.host).then(() => {
      // 要求 TCP 已经连接
      if (this.conn.isReady) {
        return this.conn.send(cmdCfg.Commands.CMPP_CONNECT, {
          Source_Addr: spId,
          AuthenticatorSource: _this.getAuthenticatorSource(spId, secret, timestamp),
          // 对于 3.0 版本 CMPP，要求 `CMPP_CONNECT` 消息和 `CMPP_CONNECT_RESP` 消息中的 `Version` 字段表示 3.0 版本;
          Version: _this.config.version,
          Timestamp: parseInt(timestamp)
        }).then(() => {
          _this.conn.cmppConnected = true; // 鉴权通过
        }).catch((err) => {
          _this.debug('cmpp_connect failed')
          _this.conn.destroySocket(_this.conn.socket)
          throw err;
        });
      }
    })
  }

  disconnect() {
    if (!this.conn.isReady) return Promise.resolve()
    return this.conn.disconnect()
  }

  /**
   * 设置上行回调处理方法
   * @param {!function} receive
   */
  setReceive(receive) {
    this.receive = receive;
  }

  /**
   * 设置下行回调处理方法
   * @param {!function} deliver
   */
  setDeliver(deliver) {
    this.deliver = deliver;
  }

  bindEvent() {
    const _this = this
    /**
     * @param {!object} data - 回调
     * @param {!function} next - 发送该 deliver 相对的 CMPP_DELIVER_RESP 响应. sendRes(Result = 0) 其中参数 Result 为响应中 Result.
     */
    this.conn.on('deliver', (data, sendRes) => {
      const body = data.body
      let handler;
      switch (body.Registered_Delivery) {
        case 1:
          handler = _this.deliver(body);
          break
        default:
          handler = _this.receive(body);
      }

      handler = isPromise(handler) ? handler : Promise.resolve(handler);
      let done = false;
      const once = (...args) => {
        if (!done) {
          done = true;
          sendRes(...args);
        }
      }
      // 先处理 deliver 再响应
      handler.then(() => {
        once();
      }).catch((err) => {
        // 7:业务代码错
        once(7);
      })
    })

    this.conn.on('terminated', (message) => {
        _this.emit('terminated', message)
    })

    this.conn.on('error', (err) => {
        _this.emit('error', err)
    })

    this.conn.on('timeout', (command) => {
      _this.emit('timeout', command)
    })

    this.conn.on('connect', () => {
      _this.emit('connect')
    })
  }

  sendGroup(mobileList, content, extendCode) {
    if (!this.conn.isReady) {
      return Promise.reject(new Error(`socket is not Ready. ${RetryInfo}`))
    }

    // CMPP 连接前无法发送
    if (!this.conn.cmppConnected) {
      return Promise.reject(new Error(`cmpp connect is not Ready. ${RetryInfo}`));
    }

    if ((this.sendingMobileCount + mobileList.length) > this.config.mobilesPerSecond) {
        return Promise.reject(new Error(`cmpp exceed max mobilesPerSecond[${this.config.mobilesPerSecond}], ${RetryInfo}`))
    }
    this.sendingMobileCount += mobileList.length
    const body = this.buildSubmitBody()
    const destBuffer = Buffer.alloc(mobileList.length * 21, 0)
    mobileList.forEach((mobile, index) => {
        destBuffer.write(mobile, index * 21, 21, 'ascii')
    })
    body.DestUsr_tl = mobileList.length
    body.Dest_terminal_Id = destBuffer
    if (extendCode) {
      body.Src_Id = body.Src_Id + extendCode;
    }
    if (content.length > this.contentLimit) {
        return this.sendLongSms(body, content)
    }

    const str = iconv.encode(content, 'gbk')
    const buf = Buffer.from(str)
    body.Msg_Length = buf.length
    body.Msg_Content = buf

    return this.conn.send(cmdCfg.Commands.CMPP_SUBMIT, body)
  }

  sendLongSms(body, content) {
    const _this = this
    const buf = Buffer.from(content, 'ucs2').swap16() // big-endian
    const bufSliceCount = this.longSmsBufLimit - 6
    const splitCount = Math.ceil(buf.length / bufSliceCount)
    const tp_udhiHead_buf = Buffer.alloc(6)
    tp_udhiHead_buf[0] = 5
    tp_udhiHead_buf[1] = 0
    tp_udhiHead_buf[2] = 3
    tp_udhiHead_buf[3] = _.random(127)
    tp_udhiHead_buf[4] = splitCount
    const promiseList = []
    _.times(splitCount, function (idx) {
        tp_udhiHead_buf[5] = idx + 1
        body.TP_udhi = 1
        body.Msg_Fmt = 8
        body.Pk_total = splitCount
        body.Pk_number = idx + 1
        body.Msg_Content = Buffer.concat([tp_udhiHead_buf, buf.slice(bufSliceCount * idx, bufSliceCount * (idx + 1))])
        body.Msg_Length = body.Msg_Content['length']
        promiseList.push(_this.conn.send(cmdCfg.Commands.CMPP_SUBMIT, body))
    })
    return Promise.all(promiseList)
  }

  send(mobile, content, extendCode) {
    return this.sendGroup([mobile], content, extendCode)
  }

  buildSubmitBody() {
    return  this.config.version === 0x30 ? {
      Msg_Id: '',
      Pk_total: 1,
      Pk_number: 1,
      Registered_Delivery: 1,
      Msg_level: 1,
      Service_Id: this.config.serviceId,
      Fee_UserType: 2,
      Fee_terminal_Id: '',
      Fee_terminal_type: 1, // 0:真实号码;1: 伪码
      TP_pId: 0,
      TP_udhi: 0,
      Msg_Fmt: 15,
      Msg_src: this.spId,
      FeeType: '03',
      FeeCode: this.config.feeCode,
      ValId_Time: '',
      At_Time: '',
      Src_Id: this.config.srcId,
      DestUsr_tl: 1,
      Dest_terminal_Id: '',
      Dest_terminal_type: 0, // 0:真实号码;1: 伪码
      Msg_Length: 0,
      Msg_Content: '',
      // Reserve: ''
      LinkID: ''
    } : {
      Msg_Id: '',
      Pk_total: 1,
      Pk_number: 1,
      Registered_Delivery: 1,
      Msg_level: 1,
      Service_Id: this.config.serviceId,
      Fee_UserType: 2,
      Fee_terminal_Id: '',
      //Fee_terminal_type: 1,
      TP_pId: 0,
      TP_udhi: 0,
      Msg_Fmt: 15,
      Msg_src: this.spId,
      FeeType: '03',
      FeeCode: this.config.feeCode,
      ValId_Time: '',
      At_Time: '',
      Src_Id: this.config.srcId,
      DestUsr_tl: 1,
      Dest_terminal_Id: '',
      Dest_terminal_type: 0,
      Msg_Length: 0,
      Msg_Content: '',
      Reserve: ''
      //LinkID: ''
    }
  }

  clearMobileCount() {
    const _this = this
    this.emit('mobile_count', this.sendingMobileCount)
    if (this.sendingMobileCount >= this.config.mpsThreshold * this.config.mobilesPerSecond) {
      this.emit('overflow', this.sendingMobileCount);
    }
    this.sendingMobileCount = 0
    setTimeout(() => {
      _this.clearMobileCount()
    }, 1000)
  }

  getAuthenticatorSource(spId, secret, timestamp) {
    const buffers = []
    buffers.push(Buffer.from(spId, 'ascii'));
    buffers.push(Buffer.alloc(9, 0));
    buffers.push(Buffer.from(secret, 'ascii'));
    buffers.push(Buffer.from(timestamp, 'ascii'));
    const buffer = Buffer.concat(buffers);

    return require('crypto').createHash('md5').update(buffer).digest();
  }

  getTimestamp() {
    return moment().format('MMDDHHmmss')
  }
}

module.exports.Client = Client


lib/commandsConfig.js
(function (Commands) {
  Commands[Commands['CMPP_CONNECT'] = 0x00000001] = 'CMPP_CONNECT'
  Commands[Commands['CMPP_CONNECT_RESP'] = 0x80000001] = 'CMPP_CONNECT_RESP'
  Commands[Commands['CMPP_SUBMIT'] = 0x00000004] = 'CMPP_SUBMIT'
  Commands[Commands['CMPP_SUBMIT_RESP'] = 0x80000004] = 'CMPP_SUBMIT_RESP'
  Commands[Commands['CMPP_DELIVER'] = 0x00000005] = 'CMPP_DELIVER'
  Commands[Commands['CMPP_DELIVER_RESP'] = 0x80000005] = 'CMPP_DELIVER_RESP'
  Commands[Commands['CMPP_ACTIVE_TEST'] = 0x00000008] = 'CMPP_ACTIVE_TEST'
  Commands[Commands['CMPP_ACTIVE_TEST_RESP'] = 0x80000008] = 'CMPP_ACTIVE_TEST_RESP'
  Commands[Commands['CMPP_TERMINATE'] = 0x00000002] = 'CMPP_TERMINATE'
  Commands[Commands['CMPP_TERMINATE_RESP'] = 0x80000002] = 'CMPP_TERMINATE_RESP'
})(exports.Commands || (exports.Commands = {}))
let Commands = exports.Commands;

(function (Errors) {
  Errors[Errors['消息结构错'] = 1] = '消息结构错'
  Errors[Errors['命令字错误'] = 2] = '命令字错误'
  Errors[Errors['消息序列号重复'] = 3] = '消息序列号重复'
  Errors[Errors['消息长度错'] = 4] = '消息长度错'
  Errors[Errors['资费代码错'] = 5] = '资费代码错'
  Errors[Errors['超过最大信息长'] = 6] = '超过最大信息长'
  Errors[Errors['业务代码错'] = 7] = '业务代码错'
  Errors[Errors['流量控制错'] = 8] = '流量控制错'
  Errors[Errors['本网关不负责此计费号码'] = 9] = '本网关不负责此计费号码'
  Errors[Errors['Src_ID错'] = 10] = 'Src_ID错'
  Errors[Errors['Msg_src错'] = 11] = 'Msg_src错'
  Errors[Errors['计费地址错'] = 12] = '计费地址错'
  Errors[Errors['目的地址错'] = 13] = '目的地址错'
  Errors[Errors['尚未建立连接'] = 51] = '尚未建立连接'
  Errors[Errors['尚未成功登录'] = 52] = '尚未成功登录'
  Errors[Errors['发送消息失败'] = 53] = '发送消息失败'
  Errors[Errors['超时未接收到响应消息'] = 54] = '超时未接收到响应消息'
  Errors[Errors['等待状态报告超时'] = 55] = '等待状态报告超时'
  Errors[Errors['有效时间已经过期'] = 61] = '有效时间已经过期'
  Errors[Errors['定时发送时间已经过期'] = 62] = '定时发送时间已经过期'
  Errors[Errors['不能识别的FeeType'] = 63] = '不能识别的FeeType'
  Errors[Errors['发送服务源地址鉴权失败'] = 64] = '发送服务源地址鉴权失败'
  Errors[Errors['发送服务目的地址鉴权失败'] = 65] = '发送服务目的地址鉴权失败'
  Errors[Errors['接收服务源地址鉴权失败'] = 66] = '接收服务源地址鉴权失败'
  Errors[Errors['接收服务目的地址鉴权失败'] = 67] = '接收服务目的地址鉴权失败'
  Errors[Errors['用户鉴权失败'] = 68] = '用户鉴权失败'
  Errors[Errors['此用户为黑名单用户'] = 69] = '此用户为黑名单用户'
  Errors[Errors['网络断连或目的设备关闭接口'] = 70] = '网络断连或目的设备关闭接口'
  Errors[Errors['超过最大节点数'] = 71] = '超过最大节点数'
  Errors[Errors['找不到路由'] = 72] = '找不到路由'
  Errors[Errors['等待应答超时'] = 73] = '等待应答超时'
  Errors[Errors['送SCP失败'] = 74] = '送SCP失败'
  Errors[Errors['送SCP鉴权等待应答超时'] = 75] = '送SCP鉴权等待应答超时'
  Errors[Errors['信息安全鉴权失败'] = 76] = '信息安全鉴权失败'
  Errors[Errors['超过最大Submit提交数'] = 77] = '超过最大Submit提交数'
  Errors[Errors['SPID为空'] = 78] = 'SPID为空'
  Errors[Errors['业务类型为空'] = 79] = '业务类型为空'
  Errors[Errors['CPCode错误'] = 80] = 'CPCode错误'
  Errors[Errors['发送接收接口重复'] = 81] = '发送接收接口重复'
  Errors[Errors['循环路由'] = 82] = '循环路由'
  Errors[Errors['超过接收侧短消息MTU'] = 83] = '超过接收侧短消息MTU'
  Errors[Errors['送DSMP重发失败'] = 84] = '送DSMP重发失败'
  Errors[Errors['DSMP系统忙重发'] = 85] = 'DSMP系统忙重发'
  Errors[Errors['DSMP系统忙且缓存满重发'] = 86] = 'DSMP系统忙且缓存满重发'
  Errors[Errors['DSMP流控重发'] = 87] = 'DSMP流控重发'
  Errors[Errors['等DSMP应答超时重发'] = 88] = '等DSMP应答超时重发'
  Errors[Errors['非神州行预付费用户'] = 202] = '非神州行预付费用户'
  Errors[Errors['数据库操作失败'] = 203] = '数据库操作失败'
  Errors[Errors['移动用户帐户数据异常'] = 206] = '移动用户帐户数据异常'
  Errors[Errors['用户余额不足'] = 208] = '用户余额不足'
  Errors[Errors['超过最高欠费额'] = 210] = '超过最高欠费额'
  Errors[Errors['重复发送消息序列号msgid相同的计费请求消息'] = 215] = '重复发送消息序列号msgid相同的计费请求消息'
  Errors[Errors['SCP互联失败'] = 218] = 'SCP互联失败'
  Errors[Errors['未登记的SP'] = 222] = '未登记的SP'
  Errors[Errors['月消费超额'] = 232] = '月消费超额'
  Errors[Errors['未定义'] = 241] = '未定义'
  Errors[Errors['消息队列满'] = 250] = '消息队列满'
})(exports.Errors || (exports.Errors = {}))
const Errors = exports.Errors;

(function (Status) {
  Status[Status['消息结构错'] = 1] = '消息结构错'
  Status[Status['非法源地址'] = 2] = '非法源地址'
  Status[Status['认证错'] = 3] = '认证错'
  Status[Status['版本太高'] = 4] = '版本太高'
  Status[Status['超过系统接口数'] = 55] = '超过系统接口数'
  Status[Status['超过帐号设置接口数'] = 56] = '超过帐号设置接口数'
  Status[Status['SP登陆IP错误'] = 57] = 'SP登陆IP错误'
  Status[Status['创建soap处理线程失败'] = 58] = '创建soap处理线程失败'
  Status[Status['登陆帐号并非属于登陆的PROXY'] = 60] = '登陆帐号并非属于登陆的PROXY'
})(exports.Status || (exports.Status = {}))
const Status = exports.Status;
exports.CommandsDescriptionV2 = {
  CMPP_CONNECT: [
      { name: 'Source_Addr', type: 'string', length: 6 },
      { name: 'AuthenticatorSource', type: 'buffer', length: 16 },
      { name: 'Version', type: 'number', length: 1 },
      { name: 'Timestamp', type: 'number', length: 4 }
  ],
  CMPP_CONNECT_RESP: [
      { name: 'Status', type: 'number', length: 1 },
      { name: 'AuthenticatorISMG', type: 'buffer', length: 16 },
      { name: 'Version', type: 'number', length: 1 }
  ],
  CMPP_ACTIVE_TEST_RESP: [
    { name: 'Reserved', type: 'number', length: 1 },
  ],
  CMPP_SUBMIT: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Pk_total', type: 'number', length: 1 },
      { name: 'Pk_number', type: 'number', length: 1 },
      { name: 'Registered_Delivery', type: 'number', length: 1 },
      { name: 'Msg_level', type: 'number', length: 1 },
      { name: 'Service_Id', type: 'string', length: 10 },
      { name: 'Fee_UserType', type: 'number', length: 1 },
      // CMPP 2.0 协议中为数字, 不好编码，故改成字符串（注：CMPP 3.0 已经调整为 32 位字符串）
      { name: 'Fee_terminal_Id', type: 'string', length: 21 },
      //{ name: 'Fee_terminal_type', type: 'number', length: 1 },
      { name: 'TP_pId', type: 'number', length: 1 },
      { name: 'TP_udhi', type: 'number', length: 1 },
      { name: 'Msg_Fmt', type: 'number', length: 1 },
      { name: 'Msg_src', type: 'string', length: 6 },
      { name: 'FeeType', type: 'string', length: 2 },
      { name: 'FeeCode', type: 'string', length: 6 },
      { name: 'ValId_Time', type: 'string', length: 17 },
      { name: 'At_Time', type: 'string', length: 17 },
      { name: 'Src_Id', type: 'string', length: 21 },
      { name: 'DestUsr_tl', type: 'number', length: 1 } // < 100
      ,
      { name: 'Dest_terminal_Id', type: 'string', length: function (obj) { return obj.DestUsr_tl * 21 } },
      //{ name: 'Dest_terminal_type', type: 'number', length: 1 },
      { name: 'Msg_Length', type: 'number', length: 1 } //<= 140
      ,
      { name: 'Msg_Content', type: 'buffer', length: function (obj) { return obj.Msg_Length } },
      //{ name: 'LinkID', type: 'string', length: 20 } //留空，点播业务使用的LinkID
      { name: 'Reserve', type: 'string', length: 8 } //保留
  ],
  CMPP_SUBMIT_RESP: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Result', type: 'number', length: 1 }
  ],
  CMPP_DELIVER: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Dest_Id', type: 'string', length: 21 },
      { name: 'Service_Id', type: 'string', length: 10 },
      { name: 'TP_pid', type: 'number', length: 1 },
      { name: 'TP_udhi', type: 'number', length: 1 },
      { name: 'Msg_Fmt', type: 'number', length: 1 },
      { name: 'Src_terminal_Id', type: 'string', length: 21 },
      //{ name: 'Src_terminal_type', type: 'number', length: 1 },
      { name: 'Registered_Delivery', type: 'number', length: 1 } //0 非状态报告 1 状态报告
      ,
      { name: 'Msg_Length', type: 'number', length: 1 },
      { name: 'Msg_Content', type: 'buffer', length: function (obj) { return obj.Msg_Length } },
      { name: 'Reserve', type: 'string', length: 8 } //保留
      //{ name: 'LinkID', type: 'string', length: 20 }
  ],
  CMPP_DELIVER_REPORT_CONTENT: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Stat', type: 'string', length: 7 },
      { name: 'Submit_time', type: 'string', length: 10 },
      { name: 'Done_time', type: 'string', length: 10 },
      { name: 'Dest_terminal_Id', type: 'string', length: 21 },
      { name: 'SMSC_sequence', type: 'number', length: 4 }
  ],
  CMPP_DELIVER_RESP: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Result', type: 'number', length: 1 }
  ]
}

exports.CommandsDescriptionV3 = {
  CMPP_CONNECT: [
      { name: 'Source_Addr', type: 'string', length: 6 },
      { name: 'AuthenticatorSource', type: 'buffer', length: 16 },
      { name: 'Version', type: 'number', length: 1 },
      { name: 'Timestamp', type: 'number', length: 4 }
  ],
  CMPP_CONNECT_RESP: [
      // 把所有响应消息中的状态码字段从1个字节扩展 为4个字节
      { name: 'Status', type: 'number', length: 4 },
      { name: 'AuthenticatorISMG', type: 'buffer', length: 16 },
      { name: 'Version', type: 'number', length: 1 }
  ],
  CMPP_ACTIVE_TEST_RESP: [
    { name: 'Reserved', type: 'number', length: 1 },
  ],
  CMPP_SUBMIT: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Pk_total', type: 'number', length: 1 },
      { name: 'Pk_number', type: 'number', length: 1 },
      { name: 'Registered_Delivery', type: 'number', length: 1 },
      { name: 'Msg_level', type: 'number', length: 1 },
      { name: 'Service_Id', type: 'string', length: 10 },
      { name: 'Fee_UserType', type: 'number', length: 1 },
      // CMPP_SUBMIT消息:扩展Fee_terminal_Id长度 为32字节，适应伪码的长度需求，并把其类型从Unsigned Integer修改为Octet String。
      { name: 'Fee_terminal_Id', type: 'string', length: 32 },
      // CMPP_SUBMIT消息:增加Fee_terminal_type 字段，表明Fee_terminal_Id是真实用户号码还是伪码
      { name: 'Fee_terminal_type', type: 'number', length: 1 },
      { name: 'TP_pId', type: 'number', length: 1 },
      { name: 'TP_udhi', type: 'number', length: 1 },
      { name: 'Msg_Fmt', type: 'number', length: 1 },
      { name: 'Msg_src', type: 'string', length: 6 },
      { name: 'FeeType', type: 'string', length: 2 },
      { name: 'FeeCode', type: 'string', length: 6 },
      { name: 'ValId_Time', type: 'string', length: 17 },
      { name: 'At_Time', type: 'string', length: 17 },
      { name: 'Src_Id', type: 'string', length: 21 },
      { name: 'DestUsr_tl', type: 'number', length: 1 }, // < 100
      // CMPP_SUBMIT消息:扩展Dest_terminal_Id的 单元长度为32字节，适应伪码的长度需求
      { name: 'Dest_terminal_Id', type: 'string', length: function (obj) { return obj.DestUsr_tl * 32 } },
      // CMPP_SUBMIT消息:增加Dest_terminal_type 字段，表明Dest_terminal_Id是真实用户号码还是伪码
      { name: 'Dest_terminal_type', type: 'number', length: 1 },
      { name: 'Msg_Length', type: 'number', length: 1 } //<= 140
      ,
      { name: 'Msg_Content', type: 'buffer', length: function (obj) { return obj.Msg_Length } },
      // 删除CMPP_SUBMIT、CMPP_DELIVER、 CMPP_FWD消息中的Reserve字段，添加LinkID字 段;(20个字节长字符串类型);
      { name: 'LinkID', type: 'string', length: 20 } //留空，点播业务使用的LinkID
      // { name: 'Reserve', type: 'string', length: 8 } //保留
  ],
  CMPP_SUBMIT_RESP: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      // 把所有响应消息中的状态码字段从1个字节扩展 为4个字节
      { name: 'Result', type: 'number', length: 4 }
  ],
  CMPP_DELIVER: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Dest_Id', type: 'string', length: 21 },
      { name: 'Service_Id', type: 'string', length: 10 },
      { name: 'TP_pid', type: 'number', length: 1 },
      { name: 'TP_udhi', type: 'number', length: 1 },
      { name: 'Msg_Fmt', type: 'number', length: 1 },
      { name: 'Src_terminal_Id', type: 'string', length: 32 },
      // CMPP_DELIVER消息:增加Src_terminal_type 字段，表明Src_terminal_Id是真实用户号码还是伪码;
      { name: 'Src_terminal_type', type: 'number', length: 1 },
      { name: 'Registered_Delivery', type: 'number', length: 1 } //0 非状态报告 1 状态报告
      ,
      { name: 'Msg_Length', type: 'number', length: 1 },
      { name: 'Msg_Content', type: 'buffer', length: function (obj) { return obj.Msg_Length } },
      // 删除CMPP_SUBMIT、CMPP_DELIVER、 CMPP_FWD消息中的Reserve字段，添加LinkID字 段;(20个字节长字符串类型);
      // { name: 'Reserve', type: 'string', length: 8 } //保留
      { name: 'LinkID', type: 'string', length: 20 }
  ],
  CMPP_DELIVER_REPORT_CONTENT: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      { name: 'Stat', type: 'string', length: 7 },
      { name: 'Submit_time', type: 'string', length: 10 },
      { name: 'Done_time', type: 'string', length: 10 },
      // CMPP_DELIVER消息传递内容为状态报告时， Msg_Content中的Dest_terminal_Id字段的长度从21 个字节扩充为32个字节;
      { name: 'Dest_terminal_Id', type: 'string', length: 32 },
      { name: 'SMSC_sequence', type: 'number', length: 4 }
  ],
  CMPP_DELIVER_RESP: [
      { name: 'Msg_Id', type: 'buffer', length: 8 },
      // 把所有响应消息中的状态码字段从1个字节扩展 为4个字节
      { name: 'Result', type: 'number', length: 4 }
  ]
}